### PR TITLE
Zh export issues

### PR DESCRIPTION
**Pull request by Tarcontar**

### CODE REVIEW

**Comment by feliwir** [comment]
> from now on only bones should be animated, test if changing their visibility also hides associated meshes. Also what should we do on export if a user has animated a mesh instead of its bone?

So there are multiple things we need to consider:
- Hierarchical model is exported but there exists no armature --> I think then it would make sense to autogenerate the hierarchy as we did before. Not required right now, not sure what would be the best behaviour here.
- As you said: There exists an armature, but the mesh is animated instead of the bones --> Ignore the mesh animation and maybe show a warning.

Maybe we should get some user opinions

File: io_mesh_w3d/export_utils.py [hunk 1]
```diff
@@ -40,7 +40,11 @@ def retrieve_data(context, export_settings):
         context.error('Filename is longer than ' + str(STRING_LENGTH) + ' characters, aborting export!')
         return None
 
-    (hierarchy, rig) = retrieve_hierarchy(context, container_name)
+    hierarchy, rig, hlod = None, None, None
+
+    if export_mode != 'M':
+        hierarchy, rig = retrieve_hierarchy(context, container_name)
+        hlod = create_hlod(hierarchy, container_name)
```
Comment (Tarcontar) [inline io_mesh_w3d/export_utils.py:47]: unittest -> make sure those two functions are not called on export mode M

Comment (Tarcontar) [reply io_mesh_w3d/export_utils.py:47]: not necessary

File: io_mesh_w3d/w3d/export_w3d.py [hunk 2]
```diff
@@ -25,6 +25,7 @@ def save(context, export_settings, data_context):
     elif export_mode == 'HM' or export_mode == 'HAM':
         if export_mode == 'HAM' \
                 or not export_settings['use_existing_skeleton']:
+            data_context.hlod.header.hierarchy_name = data_context.container_name
             data_context.hierarchy.header.name = data_context.container_name
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply io_mesh_w3d/w3d/export_w3d.py:29]: done

File: io_mesh_w3d/w3d/export_w3d.py [hunk 3]
```diff
@@ -39,6 +40,7 @@ def save(context, export_settings, data_context):
 
         data_context.hlod.write(file)
         if export_mode == 'HAM':
+            data_context.animation.header.hierarchy_name = data_context.container_name
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply io_mesh_w3d/w3d/export_w3d.py:43]: done

File: io_mesh_w3d/common/utils/mesh_export.py [hunk 4]
```diff
@@ -180,7 +177,7 @@ def retrieve_meshes(context, hierarchy, rig, container_name, force_vertex_materi
                     img = base_col_tex.image
                     filepath = os.path.basename(img.filepath)
                     if filepath == '':
-                        filepath = img.name + '.dds'
+                        filepath = img.name
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply io_mesh_w3d/common/utils/mesh_export.py:189]: done

File: io_mesh_w3d/common/utils/animation_export.py [hunk 5]
```diff
@@ -109,17 +109,22 @@ def retrieve_channels(obj, hierarchy, timecoded, name=None):
...
 
 
-def retrieve_animation(animation_name, hierarchy, rig, timecoded):
+def retrieve_animation(context, animation_name, hierarchy, rig, timecoded):
     channels = []
 
     for mesh in get_objects('MESH'):
-        channels.extend(retrieve_channels(mesh, hierarchy, timecoded, mesh.name))
+        if (retrieve_channels(mesh, hierarchy, timecoded, mesh.name)):
+            context.warning('Mesh \'' + mesh.name + '\' is animated, animate its parent bone instead!')
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply io_mesh_w3d/common/utils/animation_export.py:117]: done

File: io_mesh_w3d/common/utils/animation_export.py [hunk 6]
```diff
@@ -109,17 +109,22 @@ def retrieve_channels(obj, hierarchy, timecoded, name=None):
...
+    if (len(rigs) > 0):
+        channels.extend(retrieve_channels(rigs[0], hierarchy, timecoded))
 
-    for armature in bpy.data.armatures:
-        channels.extend(retrieve_channels(armature, hierarchy, timecoded))
+    if (len(bpy.data.armatures) > 0):
+        channels.extend(retrieve_channels(bpy.data.armatures[0], hierarchy, timecoded))
+
+    if (len(rigs) > 1 or len(bpy.data.armatures) > 1):
+        context.warning('Scene should only contain a single armature! -> exporting only animations of the first one: ' + rigs[0].name)
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply io_mesh_w3d/common/utils/animation_export.py:127]: done

File: io_mesh_w3d/common/utils/mesh_export.py [hunk 7]
```diff
@@ -151,6 +156,10 @@ def retrieve_meshes(context, hierarchy, rig, container_name, force_vertex_materi
         for i, material in enumerate(mesh.materials):
             mat_pass = MaterialPass()
 
+            if material is None:
+                context.warning('mesh \'' + mesh_object.name + '\' uses a invalid/empty material!')
+                continue
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply io_mesh_w3d/common/utils/mesh_export.py:162]: done

File: io_mesh_w3d/common/utils/mesh_export.py [hunk 8]
```diff
@@ -220,10 +227,13 @@ def retrieve_meshes(context, hierarchy, rig, container_name, force_vertex_materi
 def prepare_bmesh(context, mesh):
     b_mesh = bmesh.new()
     b_mesh.from_mesh(mesh)
+    bmesh.ops.triangulate(b_mesh, faces=b_mesh.faces)
+    b_mesh.to_mesh(mesh)
+    b_mesh.free()
```
[the remark this answers is not in the document]
Comment (Tarcontar) [reply io_mesh_w3d/common/utils/mesh_export.py:235]: done

**Pull request merged**